Add repo URL support and testing for cocoapods

johnmhoran opened the pull request:
Reference: https://github.com/package-url/packageurl-python/issues/143

@TG1999   This replaces the recently-closed [PR 151](https://github.com/package-url/packageurl-python/pull/151).  I look forward to your comments and questions.

TG1999 commented on tests/contrib/test_purl2url.py:
```diff
@@ -134,7 +136,10 @@ def test_purl2url_get_inferred_urls():
+        "pkg:pypi/sortedcontainers@2.4.0": [
+            "https://pypi.org/project/sortedcontainers/2.4.0/"
+        ],
+        "pkg:cocoapods/AFNetworking@4.0.1": ["https://cocoapods.org/pods/AFNetworking"],
```
Add a test case where there is no name, IMO that will fail

johnmhoran replied:
As noted in my other comments re removing the various try/except blocks I'd added, a test case with no name causes the entire test to fail, so I've removed the test case I'd previously added.

pombredanne replied:
I think this works as it is.

TG1999 commented on src/packageurl/contrib/purl2url.py:
```diff
@@ -304,6 +304,20 @@ def build_golang_repo_url(purl):
+    purl_data = PackageURL.from_string(purl)
+    name = purl_data.name
+
+    if name:
```
```suggestion
    if not name:
        return
    return f"https://cocoapods.org/pods/{name}"
```

TG1999 commented on src/packageurl/contrib/purl2url.py:
```diff
@@ -304,6 +304,20 @@ def build_golang_repo_url(purl):
+    """
+    Return a CocoaPods repo URL from the `purl` string.
+    """
+    purl_data = PackageURL.from_string(purl)
```
Please put this in a try/except block, this may fail

pombredanne replied:
@TG1999 we want to keep this function the same as other functions and not catch anything.

johnmhoran commented:
@TG1999 Thank you for your comments.  I've just pushed my responsive updates.

- I added `"pkg:cocoapods/": None,` to `test_purl2url_get_repo_url()` and added `"pkg:cocoapods/": [],` to `test_purl2url_get_inferred_urls()` as you requested to test cocoapods PURL inputs with no name.  

- Also made a few changes to your suggested try/except approach -- I added try/except blocks to both `get_download_url()` and `build_cocoapods_repo_url()`, since `get_inferred_urls()` calls both `get_repo_url` and `get_download_url` and cocoapods only has `repo_url` support in `purl2url.py` (`download_url` support is now in `fetchcode/package.py`).

Note: these changes do not handle exceptions raised by other supported PURLs with no name, e.g., `"pkg:cargo/abc": "https://crates.io/crates/abc",` in `test_purl2url_get_repo_url()` raises an exception `ValueError: purl is missing the required name component: 'pkg:cargo/'`.  

Question: since this output is consumed by the `urls` command in `purldb-toolkit/purlcli.py`, I don't want to raise an exception that will halt the underlying process of the `urls` command, e.g., a user might have submitted a list of 100 PURLs to query and does not want that to be interrupted.  In `purlcli.py` and my `fetchcode/package.py` work I've used a "soft" fail approach that records an exception message in a shared `.log` file.  I assume that's not an option here.  As a placeholder I am printing the exception message to the console, but we do not want that in the PURLCLI tools.  Do you have any suggestions?

TG1999 reviewed (approved):
LGTM!

johnmhoran commented:
@TG1999 You asked me to remove the two print statements I added to `purl2url.py`, one in `get_download_url` and the other in `build_cocoapods_repo_url`.  Is this what we want for the cocoapods tests you asked me to add?  E.g.,

```
def get_download_url(purl):
    """
    Return a download URL inferred from the `purl` string.
    """
    download_url = _get_url_from_router(download_router, purl)
    if download_url:
        return download_url

    # Fallback on the `download_url` qualifier when available.
    purl_data = None
    try:
        purl_data = PackageURL.from_string(purl)
    except Exception as e:
        # print(f"An error occurred in get_download_url(): {e}")
        return
    return purl_data.qualifiers.get("download_url", None)
```

Or are we perhaps contemplating something like this:

```
def get_download_url(purl):
    """
    Return a download URL inferred from the `purl` string.
    """
    download_url = _get_url_from_router(download_router, purl)
    if download_url:
        return download_url

    # Fallback on the `download_url` qualifier when available.
    purl_data = None
    try:
        purl_data = PackageURL.from_string(purl)
    except ValueError:
        return
    return purl_data.qualifiers.get("download_url", None)
```

pombredanne commented on src/packageurl/contrib/purl2url.py:
```diff
@@ -75,7 +76,12 @@ def get_download_url(purl):
 
     # Fallback on the `download_url` qualifier when available.
-    purl_data = PackageURL.from_string(purl)
+    purl_data = None
```
I am not sure we want to swallow an exception here. If the purl is invalid, let the error bubble up and do not catch the exception.

johnmhoran replied:
Thanks @pombredanne -- I have updated `get_download_url()` to the following (actually, this reverts the code to its state before I added a try/except):

```
def get_download_url(purl):
    """
    Return a download URL inferred from the `purl` string.
    """
    download_url = _get_url_from_router(download_router, purl)
    if download_url:
        return download_url

    # Fallback on the `download_url` qualifier when available.
    purl_data = PackageURL.from_string(purl)
    return purl_data.qualifiers.get("download_url", None)
```

I also removed `"pkg:cocoapods/": [],` from `test_purl2url_get_inferred_urls()` because otherwise the test fails w/o the try/except I've removed.

pombredanne commented on src/packageurl/contrib/purl2url.py:
```diff
@@ -304,6 +310,25 @@ def build_golang_repo_url(purl):
+    """
+    purl_data = None
+    name = None
+    try:
```
Same as above: if the purl is invalid, let the error bubble up and do not catch the exception.

johnmhoran replied:
@pombredanne I'm not sure I understand your comments but hoping this is what you suggested.  I've removed the try/except as well as this if condition:

```
    if not name:
        return
```

 and the updated `build_cocoapods_repo_url()` looks like this:

```
@repo_router.route("pkg:cocoapods/.*")
def build_cocoapods_repo_url(purl):
    """
    Return a CocoaPods repo URL from the `purl` string.
    """
    purl_data = PackageURL.from_string(purl)
    name = purl_data.name
    return name and f"https://cocoapods.org/pods/{name}"
```

I also removed `"pkg:cocoapods/": None,` from `test_purl2url_get_repo_url()` because, having removed the try/except, that test fails on `"pkg:cocoapods/": None,` (which I'd added only to test the try/except).

pombredanne reviewed (changes requested):
Thanks. Do not swallow exception silently. Best is to let them propagate as-is when the purl is not valid.

pombredanne commented on src/packageurl/contrib/purl2url.py:
```diff
@@ -24,6 +24,7 @@
 # Visit https://github.com/package-url/packageurl-python for support and
 # download.
 
+
```
Remove this

johnmhoran replied:
That comment also appears at the top of `/packageurl-python/tests/contrib/test_purl2url.py` -- I've deleted it there as well.

pombredanne replied:
I meant to remove the empty line.

johnmhoran commented on src/packageurl/contrib/purl2url.py:
```diff
@@ -304,6 +310,25 @@ def build_golang_repo_url(purl):
+
+    if not name:
+        return
+    return f"https://cocoapods.org/pods/{name}"
```
Updated as shown in the comment just above ^.

johnmhoran commented:
@pombredanne @TG1999 I've just committed and pushed my changes in response to @pombredanne's recent comments.  Please take a look when you can.

johnmhoran commented:
@pombredanne I've restored the support/download URL comments we discussed.

pombredanne reviewed (dismissed):
Thanks. Looking good. Merging.

pombredanne commented on tests/contrib/test_purl2url.py:
```diff
@@ -134,7 +136,10 @@ def test_purl2url_get_inferred_urls():
             "https://gitlab.com/tg1999/firebase",
             "https://gitlab.com/tg1999/firebase/-/archive/1a122122/firebase-1a122122.tar.gz",
         ],
-        "pkg:pypi/sortedcontainers@2.4.0": ["https://pypi.org/project/sortedcontainers/2.4.0/"],
```
Can you restore the formatting as it was? The black tests are failing

johnmhoran replied:
@pombredanne I've turned off Black (it had been set in 2 VSCode locations and turned off in just one) and restored the original format for that test entry you identified.  Committed and pushed just now.

johnmhoran commented:
@pombredanne Just confirming that this PR is ready for your re-review when you have the time.

JonoYang commented:
@johnmhoran This looks good, merging.
